cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/sync/class-sync-queue.php

```diff
@@ -240,48 +262,122 @@ public function build_queue() {
 		$ids         = $attachments->get_posts();
 		// Transform attachments.
 		$return = array(
-			'pending'    => array(),
 			'done'       => array(),
 			'processing' => array(),
 			'error'      => array(),
+			'run_status' => array(),
 		);
+		foreach ( $this->threads as $thread ) {
+			$return[ $thread ]               = array();
+			$return['run_status'][ $thread ] = array();
+		}
 
 		// Add items to pending queue.
-		$return['pending'] = $ids;
+		if ( ! empty( $ids ) ) {
+			$chunk_size = ceil( count( $ids ) / count( $this->threads ) );
+			$chunks     = array_chunk( $ids, $chunk_size );
+			foreach ( $chunks as $index => $chunk ) {
+				$return[ $this->threads[ $index ] ] = $chunk;
+			}
+		}
 
 		$this->set_queue( $return );
 
 		return $return;
 	}
 
+	/**
+	 * Maybe stop the queue.
+	 */
+	public function stop_maybe() {
+		$status = $this->get_queue_status();
+		if ( empty( $status['pending'] ) ) {
+			$this->stop_queue();
+		}
+	}
+
 	/**
 	 * Stop the queue by removing the started flag.
 	 */
 	public function stop_queue() {
 		$queue = $this->get_queue();
 		if ( ! empty( $queue['started'] ) ) {
 			unset( $queue['started'] );
+			unset( $queue['last_update'] );
 			$this->set_queue( $queue );
 		}
-		$this->running = false;
 
-		wp_clear_scheduled_hook( 'cloudinary_resume_queue' );
+		wp_unschedule_hook( 'cloudinary_resume_queue' );
 	}
 
 	/**
 	 * Start the queue by setting the started flag.
 	 *
-	 * @return void
+	 * @return array
 	 */
 	public function start_queue() {
-		$queue            = $this->get_queue();
-		$queue['started'] = $queue['last_update'] = current_time( 'timestamp' );
+		$queue = $this->get_queue();
+		if ( ! empty( $queue['processing'] ) ) {
+			// In case it stopped mid process, push back to the  first thread.
+			$queue['thread_0'] = array_merge( $queue['thread_0'], $queue['processing'] );
```

Comment:
Shouldn't we do this is all threads?

Reply:
Yes, but it got cumbersome to have to split it up when there's generally only 1 or 2 that get stuck.

Reply:
@DavidCramer 

Not sure if I follow. Aren't these threads running in parallel and independent of each other? Perhaps we can discuss this is our next tech sync?

Reply:
Did you guys figure this out in the sync?